### ISSUE TITLE
pip-sync 7.0.0 reinstalls packages due to case difference

### ISSUE THREAD

**Issue by sshishov** [jazzband/pip-tools#1918]
#### Description

pip-sync 7.0.0 reinstalls all packages which have non-lowercase letters in their names

For instance in `*.in` and `*.txt` files we have the following data:
```
django
```
and
```
django==4.2.0
  ...
```
When we run pip sync, it will uninstall and install django because in `pip-freeze` the same package mentioned as `Django`

#### Environment Versions

| Required | Information |
| ---------- | -------------- |
| OS Type | MacOs/Linux |
| Python version: `$ python -V` | Python 3.11.4 |
| pip version: `$ pip --version`: |pip 23.2 from /home/user/venv/lib/python3.11/site-packages/pip (python 3.11) |
| pip-tools version: `$ pip-compile --version` | pip-compile, version 7.0.0 |

#### Steps to replicate

1. Create `*.in` file with `django` requirement
2. Compile requirements to generate `*.txt` file
3. Run pip-sync on the txt file
4. Run pip-sync on the txt file again

#### Expected result

The output should be `Everything up-to-date`

#### Actual result

The requirement is reinstalled every time you run `pip-sync`

**Comment by atugushev** [comment]
Thanks for the issue! See https://github.com/jazzband/pip-tools/issues/1917#issuecomment-1638197374 for details.

**Comment by atugushev** [comment]
@sshishov could you test this fix #1919? Does it resolve the issue?

```
pip install git+https://github.com/atugushev/pip-tools@fix-broken-sync-diff
```

**Comment by sshishov** [comment]
@atugushev going to test it now, sorry for the long reply
